File: android/app/src/main/kotlin/com/example/count_to_three/alarm/AlarmEngine.kt
package com.example.count_to_three.alarm

import android.content.Context
import com.example.count_to_three.alarm.model.AlarmData

object AlarmEngine {

    fun init(context: Context) = AlarmStore.init(context)

    fun scheduleAlarm(
        context: Context,
        alarmId: Int,
        reminderId: String,
        title: String,
        triggerAtMs: Long,
        snoozeMinutes: Int = 5,
        maxSnoozeCount: Int = 3,
        volumeRamp: Boolean = false,
        vibrate: Boolean = true,
        ringtoneUri: String? = null,
    ) {
        val alarm = AlarmData(
            id = alarmId,
            reminderId = reminderId,
            title = title,
            scheduledAt = triggerAtMs,
            snoozeMinutes = snoozeMinutes,
            maxSnoozeCount = maxSnoozeCount,
            volumeRamp = volumeRamp,
            vibrate = vibrate,
            ringtoneUri = ringtoneUri,
        )
        AlarmStore.put(context, alarm)
        AlarmScheduler.schedule(context, alarm)
    }

    fun cancelAlarm(context: Context, alarmId: Int) {
        AlarmScheduler.cancel(context, alarmId)
        AlarmStore.remove(context, alarmId)
    }

    fun snoozeAlarm(context: Context, alarmId: Int) {
        val alarm = AlarmStore.get(alarmId) ?: return
        if (alarm.snoozeCount >= alarm.maxSnoozeCount) {
            cancelAlarm(context, alarmId)
            AlarmEventBus.emit(com.example.count_to_three.alarm.model.AlarmEvent.Dismissed(alarmId, alarm.reminderId, alarm.scheduledAt, auto = true))
            return
        }
        val snoozed = alarm.copy(
            snoozeCount = alarm.snoozeCount + 1,
            scheduledAt = System.currentTimeMillis() + alarm.snoozeMinutes * 60_000L,
        )
        AlarmStore.put(context, snoozed)
        AlarmScheduler.schedule(context, snoozed)
        AlarmEventBus.emit(com.example.count_to_three.alarm.model.AlarmEvent.Snoozed(alarmId, alarm.reminderId, alarm.scheduledAt, snoozed.snoozeCount))
    }

    fun getPendingAlarms(): List<Map<String, Any>> =
        AlarmStore.getAll().map {
            mapOf(
                "alarmId" to it.id,
                "reminderId" to it.reminderId,
                "title" to it.title,
                "triggerAtMs" to it.scheduledAt,
                "snoozeCount" to it.snoozeCount,
            )
        }
}

File: ios/Runner/Alarm/AlarmEngine.swift
import Foundation

final class AlarmEngine {
    static let shared = AlarmEngine()
    private init() {}

    // MARK: - Permission

    @discardableResult
    func requestPermission() async -> Bool {
        // Always request UNUserNotificationCenter auth (used as AlarmKit fallback on simulator)
        let notifGranted = await withCheckedContinuation { cont in
            NotificationFallback.shared.requestAuthorization { cont.resume(returning: $0) }
        }
        if #available(iOS 26.1, *) {
            let kitGranted = await AlarmKitScheduler.shared.requestAuthorization()
            return kitGranted || notifGranted
        }
        return notifGranted
    }

    // MARK: - Schedule

    func schedule(alarm: AlarmData) async throws {
        AlarmStore.shared.put(alarm)
        if #available(iOS 26.1, *) {
            do {
                try await AlarmKitScheduler.shared.schedule(alarm)
                return
            } catch {
                // AlarmKit unavailable (simulator, unauthorized) — fall through to UNNotification
            }
        }
        NotificationFallback.shared.registerCategory()
        NotificationFallback.shared.schedule(alarm)
    }

    // MARK: - Snooze

    func snooze(alarmId: Int) async throws {
        guard var alarm = AlarmStore.shared.get(alarmId) else { return }

        let originalScheduledAt = alarm.scheduledAt

        if alarm.snoozeCount >= alarm.maxSnoozeCount {
            // Max snoozes reached — just dismiss
            try await cancel(alarmId: alarmId)
            AlarmEventBus.shared.emit([
                "type": "dismissed",
                "alarmId": alarmId,
                "reminderId": alarm.reminderId,
                "scheduledAtMs": originalScheduledAt,
                "auto": true,
            ])
            return
        }

        // Cancel current alarm
        if #available(iOS 26.1, *) { try? AlarmKitScheduler.shared.cancel(alarmId: alarmId) }
        NotificationFallback.shared.cancel(alarmId: alarmId)

        // Reschedule — schedule(alarm:) calls AlarmStore.shared.put internally.
        alarm.snoozeCount += 1
        alarm.scheduledAt = Int(Date().timeIntervalSince1970 * 1000) + alarm.snoozeMinutes * 60_000

        try await schedule(alarm: alarm)

        AlarmEventBus.shared.emit([
            "type": "snoozed",
            "alarmId": alarmId,
            "reminderId": alarm.reminderId,
            "scheduledAtMs": originalScheduledAt,
            "snoozeCount": alarm.snoozeCount,
        ])
    }

    // MARK: - Cancel

    func cancel(alarmId: Int) async throws {
        if #available(iOS 26.1, *) {
            try? AlarmKitScheduler.shared.cancel(alarmId: alarmId)
        }
        // Also cancel UNNotification in case this alarm used the fallback path
        NotificationFallback.shared.cancel(alarmId: alarmId)
        AlarmStore.shared.remove(alarmId)
    }

    // MARK: - Foreground reconciliation (Decision C)

    /// Called from applicationDidBecomeActive.
    /// Compares AlarmStore vs AlarmKit's live list; past-due alarms that vanished
    /// were stopped/snoozed via the system UI — emit Dismissed events and clean up.
    func detectHandledAlarms() async {
        guard #available(iOS 26.1, *) else { return }
        guard let activeIds = await AlarmKitScheduler.shared.pendingAlarmIds() else { return }
        let nowMs = Int(Date().timeIntervalSince1970 * 1000)
        for alarm in AlarmStore.shared.getAll() {
            // Only reconcile alarms that should have already fired; future
            // alarms won't be in the live list yet even when correctly scheduled.
            guard alarm.scheduledAt <= nowMs else { continue }
            guard !activeIds.contains(alarm.id) else { continue }
            AlarmEventBus.shared.emit([
                "type": "dismissed",
                "alarmId": alarm.id,
                "reminderId": alarm.reminderId,
                "scheduledAtMs": alarm.scheduledAt,
                "auto": false,
            ])
            AlarmStore.shared.remove(alarm.id)
        }
    }
}
